maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,115 +120,109 @@ public fun GoogleMap(
-
-                CompositionLocalProvider(
-                    LocalCameraPositionState provides currentCameraPositionState,
-                ) {
-                    currentContent?.invoke()
-                }
-            }
-        }
-    }
-}
 
-internal suspend inline fun disposingComposition(factory: () -> Composition) {
-    val composition = factory()
-    try {
-        awaitCancellation()
-    } finally {
-        composition.dispose()
-    }
-}
+    /**
+     * Create and apply the [content] compositions to the map +
+     * dispose the [Composition] when the parent composable is disposed.
+     * */
+    fun CoroutineScope.launchSubcomposition(mapView: MapView): Job {
+        val mapCompositionContent: @Composable () -> Unit = {
+            MapUpdater(
+                mergeDescendants = mergeDescendants,
+                contentDescription = currentContentDescription,
+                cameraPositionState = currentCameraPositionState,
+                contentPadding = currentContentPadding,
+                locationSource = currentLocationSource,
+                mapProperties = currentMapProperties,
+                mapUiSettings = currentUiSettings,
+            )
 
-private suspend inline fun MapView.newComposition(
-    parent: CompositionContext,
-    mapClickListeners: MapClickListeners,
-    noinline content: @Composable () -> Unit
-): Composition {
-    val map = awaitMap()
-    return Composition(
-        MapApplier(map, this, mapClickListeners), parent
-    ).apply {
-        setContent(content)
-    }
-}
+            MapClickListenerUpdater()
 
-/**
- * Registers lifecycle observers to the local [MapView].
- */
-@Composable
-private fun MapLifecycle(mapView: MapView) {
-    val context = LocalContext.current
-    val lifecycle = LocalLifecycleOwner.current.lifecycle
-    val previousState = remember { mutableStateOf(Lifecycle.Event.ON_CREATE) }
-    DisposableEffect(context, lifecycle, mapView) {
-        val mapLifecycleObserver = mapView.lifecycleObserver(previousState)
-        val callbacks = mapView.componentCallbacks()
-
-        lifecycle.addObserver(mapLifecycleObserver)
-        context.registerComponentCallbacks(callbacks)
-
-        onDispose {
-            lifecycle.removeObserver(mapLifecycleObserver)
-            context.unregisterComponentCallbacks(callbacks)
+            CompositionLocalProvider(
+                LocalCameraPositionState provides currentCameraPositionState,
+            ) {
+                currentContent?.invoke()
+            }
         }
-    }
-    DisposableEffect(mapView) {
-        onDispose {
-            mapView.onDestroy()
-            mapView.removeAllViews()
+
+        // Use [CoroutineStart.UNDISPATCHED] to kick off GoogleMap loading immediately
+        return launch(start = CoroutineStart.UNDISPATCHED) {
+            val map = mapView.awaitMap()
+            val composition = Composition(
+                applier = MapApplier(map, mapView, mapClickListeners),
```

Comment:
`mapClickListeners` should be moved out of `MapApplier`, and passed as a parameter to `MapClickListenerUpdater()`. @philip-segerfast determined earlier in the review process that `MapClickListeners` does not actually have to be a singleton. (I verified the finding.)

 It's a trivial change that will also prepare `MapApplier` for potential future use in `ReusableComposition`.

Reply:
I've implemented this. Please tell me what you think.

Reply:
@philip-segerfast I've realized that the MapClickListeners change that I proposed is not valid, sadly. It will work in this case, because MapClickListeners actually is a singleton within the scope of a non-reused composition anyway. However, from the perspective of `MapClickListenerUpdater(MapClickListeners)` as a standalone Composable function it is not correct, as passing a different `MapClickListener` on recomposition will not do the right thing. The initial listener is too baked into the lambdas there.

Let's undo this part of the commit and leave it for the future, it does not look trivial. My sincere apologies!

Reply:
No worries. This has been reverted.